File: backend/internal/api/tamagotchi.go
package api

import (
	"encoding/json"
	"net/http"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/martbul/p-ink/internal/db"
	"github.com/martbul/p-ink/internal/errs"
	"github.com/martbul/p-ink/internal/middleware"
	"github.com/martbul/p-ink/internal/models"
)

// ─── GET /api/tamagotchi/mine ─────────────────────────────────────────────────

func GetMyTamagotchi(pool *pgxpool.Pool) http.HandlerFunc {
	const op = errs.Op("api.GetMyTamagotchi")
	return func(w http.ResponseWriter, r *http.Request) {
		user := middleware.UserFromContext(r.Context())

		tama, err := db.GetTamagotchiByOwner(r.Context(), pool, user.ID)
		if err != nil {
			Error(w, errs.E(op, errs.KindInternal, err, "failed to look up tamagotchi"))
			return
		}
		if tama == nil {
			NotFound(w, "tamagotchi not found — is your couple active?")
			return
		}

		state, err := db.GetTamagotchiState(r.Context(), pool, tama.ID)
		if err != nil {
			Error(w, errs.E(op, errs.KindInternal, err, "failed to fetch tamagotchi state"))
			return
		}
		OK(w, state)
	}
}

// ─── GET /api/tamagotchi/partner ──────────────────────────────────────────────

func GetPartnerTamagotchi(pool *pgxpool.Pool) http.HandlerFunc {
	const op = errs.Op("api.GetPartnerTamagotchi")
	return func(w http.ResponseWriter, r *http.Request) {
		user := middleware.UserFromContext(r.Context())

		tama, err := db.GetTamagotchiByController(r.Context(), pool, user.ID)
		if err != nil {
			Error(w, errs.E(op, errs.KindInternal, err, "failed to look up partner tamagotchi"))
			return
		}
		if tama == nil {
			NotFound(w, "tamagotchi not found — is your couple active?")
			return
		}

		state, err := db.GetTamagotchiState(r.Context(), pool, tama.ID)
		if err != nil {
			Error(w, errs.E(op, errs.KindInternal, err, "failed to fetch partner tamagotchi state"))
			return
		}
		OK(w, state)
	}
}

// ─── POST /api/tamagotchi/rename ──────────────────────────────────────────────

func RenameTamagotchi(pool *pgxpool.Pool) http.HandlerFunc {
	const op = errs.Op("api.RenameTamagotchi")
	return func(w http.ResponseWriter, r *http.Request) {
		user := middleware.UserFromContext(r.Context())

		var body struct {
			Name string `json:"name"`
		}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil || body.Name == "" {
			BadRequest(w, "name is required")
			return
		}
		if len(body.Name) > 32 {
			BadRequest(w, "name must be 32 characters or fewer")
			return
		}

		tama, err := db.GetTamagotchiByOwner(r.Context(), pool, user.ID)
		if err != nil {
			Error(w, errs.E(op, errs.KindInternal, err, "failed to look up tamagotchi"))
			return
		}
		if tama == nil {
			NotFound(w, "tamagotchi not found")
			return
		}

		if err := db.RenameTamagotchi(r.Context(), pool, tama.ID, body.Name); err != nil {
			Error(w, errs.E(op, errs.KindInternal, err, "failed to rename tamagotchi"))
			return
		}
		tama.Name = body.Name
		OK(w, tama)
	}
}

// ─── GET /api/tamagotchi/shop ─────────────────────────────────────────────────

func GetShop(pool *pgxpool.Pool) http.HandlerFunc {
	const op = errs.Op("api.GetShop")
	return func(w http.ResponseWriter, r *http.Request) {
		user := middleware.UserFromContext(r.Context())

		tama, err := db.GetTamagotchiByController(r.Context(), pool, user.ID)
		if err != nil {
			Error(w, errs.E(op, errs.KindInternal, err, "failed to look up tamagotchi"))
			return
		}
		if tama == nil {
			NotFound(w, "tamagotchi not found")
			return
		}

		items, err := db.ListShopItems(r.Context(), pool, tama.ID, tama.Level)
		if err != nil {
			Error(w, errs.E(op, errs.KindInternal, err, "failed to list shop items"))
			return
		}

		OK(w, map[string]any{
			"tamagotchi_xp":    tama.XP,
			"tamagotchi_level": tama.Level,
			"items":            items,
		})
	}
}

// ─── POST /api/tamagotchi/shop/buy ────────────────────────────────────────────

func BuyItem(pool *pgxpool.Pool) http.HandlerFunc {
	const op = errs.Op("api.BuyItem")
	return func(w http.ResponseWriter, r *http.Request) {
		user := middleware.UserFromContext(r.Context())

		var body struct {
			ItemID string `json:"item_id"`
		}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil || body.ItemID == "" {
			BadRequest(w, "item_id is required")
			return
		}
		itemID, err := uuid.Parse(body.ItemID)
		if err != nil {
			BadRequest(w, "invalid item_id")
			return
		}

		tama, err := db.GetTamagotchiByController(r.Context(), pool, user.ID)
		if err != nil {
			Error(w, errs.E(op, errs.KindInternal, err, "failed to look up tamagotchi"))
			return
		}
		if tama == nil {
			NotFound(w, "tamagotchi not found")
			return
		}

		if err := db.BuyItem(r.Context(), pool, tama.ID, itemID); err != nil {
			// BuyItem returns user-readable errors (not enough XP, level req, etc.)
			Error(w, errs.E(op, errs.KindInvalid, err))
			return
		}

		state, err := db.GetTamagotchiState(r.Context(), pool, tama.ID)
		if err != nil {
			Error(w, errs.E(op, errs.KindInternal, err, "failed to fetch updated state"))
			return
		}
		Created(w, state)
	}
}

// ─── POST /api/tamagotchi/equip ───────────────────────────────────────────────

func EquipItem(pool *pgxpool.Pool) http.HandlerFunc {
	const op = errs.Op("api.EquipItem")
	return func(w http.ResponseWriter, r *http.Request) {
		user := middleware.UserFromContext(r.Context())

		var body struct {
			ItemID string `json:"item_id"`
			Slot   string `json:"slot"`
		}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			BadRequest(w, "invalid JSON")
			return
		}
		if body.ItemID == "" || body.Slot == "" {
			BadRequest(w, "item_id and slot are required")
			return
		}

		itemID, err := uuid.Parse(body.ItemID)
		if err != nil {
			BadRequest(w, "invalid item_id")
			return
		}

		slot := models.EquippedSlot(body.Slot)
		switch slot {
		case models.SlotOutfit, models.SlotAccessory, models.SlotBackground, models.SlotPosition:
		default:
			BadRequest(w, "slot must be one of: outfit, accessory, background, position")
			return
		}

		tama, err := db.GetTamagotchiByController(r.Context(), pool, user.ID)
		if err != nil {
			Error(w, errs.E(op, errs.KindInternal, err, "failed to look up tamagotchi"))
			return
		}
		if tama == nil {
			NotFound(w, "tamagotchi not found")
			return
		}

		if err := db.EquipItem(r.Context(), pool, tama.ID, itemID, slot); err != nil {
			Error(w, errs.E(op, errs.KindInvalid, err))
			return
		}

		state, err := db.GetTamagotchiState(r.Context(), pool, tama.ID)
		if err != nil {
			Error(w, errs.E(op, errs.KindInternal, err, "failed to fetch updated state"))
			return
		}
		OK(w, state)
	}
}

// ─── GET /api/tamagotchi/events ───────────────────────────────────────────────

func GetEvents(pool *pgxpool.Pool) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		user := middleware.UserFromContext(r.Context())

		myTama, _ := db.GetTamagotchiByOwner(r.Context(), pool, user.ID)
		partnerTama, _ := db.GetTamagotchiByController(r.Context(), pool, user.ID)

		type eventSet struct {
			TamagotchiID string                   `json:"tamagotchi_id"`
			Events       []models.TamagotchiEvent `json:"events"`
		}

		var result []eventSet
		if myTama != nil {
			events, _ := db.GetRecentEvents(r.Context(), pool, myTama.ID, 15)
			result = append(result, eventSet{TamagotchiID: myTama.ID.String(), Events: events})
		}
		if partnerTama != nil {
			events, _ := db.GetRecentEvents(r.Context(), pool, partnerTama.ID, 15)
			result = append(result, eventSet{TamagotchiID: partnerTama.ID.String(), Events: events})
		}

		OK(w, map[string]any{"event_sets": result})
	}
}

File: frontend/api/tamagotchi.ts
/**
 * api/tamagotchi.ts
 *
 * Typed client for every tamagotchi endpoint on the Go backend.
 * All functions require an auth token (Clerk JWT).
 *
 * Usage:
 *   import { fetchMyTamagotchi, buyItem } from "@/api/tamagotchi";
 *   const state = await fetchMyTamagotchi(token);
 */

// ─── Base ─────────────────────────────────────────────────────────────────────

const BASE = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:7111";

async function req<T>(
  token: string,
  method: "GET" | "POST" | "PATCH" | "DELETE",
  path: string,
  body?: unknown,
): Promise<T> {
  const res = await fetch(`${BASE}${path}`, {
    method,
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
    body: body !== undefined ? JSON.stringify(body) : undefined,
  });

  if (!res.ok) {
    let msg = `${method} ${path} → ${res.status}`;
    try {
      const err = await res.json();
      if (err?.error) msg = err.error;
    } catch {
      /* non-JSON error body */
    }
    throw new Error(msg);
  }

  // 204 No Content
  if (res.status === 204) return undefined as T;

  return res.json();
}

// ─── Models (mirrors backend/internal/models/models.go) ───────────────────────

export type TamagotchiMood =
  | "happy"
  | "neutral"
  | "sad"
  | "sleeping"
  | "excited";

export type TamagotchiSpecies =
  | "bear"
  | "cat"
  | "bunny"
  | "dog"
  | "shark"
  | "fox"
  | "penguin";

export type ItemType =
  | "outfit"
  | "accessory"
  | "background"
  | "animation"
  | "position";

export type EquippedSlot = "outfit" | "accessory" | "background" | "position";

export interface Tamagotchi {
  id: string;
  couple_id: string;
  owner_id: string;
  controller_id: string;
  name: string;
  species: TamagotchiSpecies;
  // free cosmetic fields (stored on the tamagotchis row)
  background: string;
  animation: string;
  position: string;
  health: number;
  max_health: number;
  xp: number;
  level: number;
  mood: TamagotchiMood;
  last_fed_at?: string;
  created_at: string;
}

export interface TamagotchiItem {
  id: string;
  type: ItemType;
  name: string;
  description: string;
  xp_cost: number;
  preview_url?: string;
  species_lock?: string;
  unlocks_at_level: number;
}

export interface TamagotchiInventoryEntry {
  id: string;
  tamagotchi_id: string;
  item: TamagotchiItem;
  purchased_at: string;
}

export interface TamagotchiEquipped {
  tamagotchi_id: string;
  slot: EquippedSlot;
  item: TamagotchiItem;
}

export type TamagotchiEventType =
  | "fed"
  | "leveled_up"
  | "item_purchased"
  | "item_equipped"
  | "mood_changed"
  | "sleeping"
  | "woke_up"
  | "decay";

export interface TamagotchiEvent {
  id: string;
  tamagotchi_id: string;
  type: TamagotchiEventType;
  xp_delta: number;
  health_delta: number;
  metadata?: Record<string, unknown>;
  created_at: string;
}

export interface TamagotchiState {
  tamagotchi: Tamagotchi;
  equipped: TamagotchiEquipped[];
  inventory: TamagotchiInventoryEntry[];
  recent_events: TamagotchiEvent[];
}

export interface ShopResponse {
  tamagotchi_id: string;
  tamagotchi_xp: number;
  tamagotchi_level: number;
  xp_this_level: number;
  xp_next_level: number;
  items: TamagotchiItem[];
}

export interface EventSet {
  tamagotchi_id: string;
  role: "mine" | "partner";
  events: TamagotchiEvent[];
}

export interface EventsResponse {
  event_sets: EventSet[];
}

// ─── API calls ────────────────────────────────────────────────────────────────

/**
 * GET /api/tamagotchi/mine
 * The Tamagotchi YOU own — lives on your frame, fed by your partner.
 */
export function fetchMyTamagotchi(token: string): Promise<TamagotchiState> {
  return req<TamagotchiState>(token, "GET", "/api/tamagotchi/mine");
}

/**
 * GET /api/tamagotchi/partner
 * Your PARTNER'S Tamagotchi — the one you feed by sending content.
 */
export function fetchPartnerTamagotchi(token: string): Promise<TamagotchiState> {
  return req<TamagotchiState>(token, "GET", "/api/tamagotchi/partner");
}

/**
 * POST /api/tamagotchi/rename
 * Rename the Tamagotchi you own.
 */
export function renameTamagotchi(token: string, name: string): Promise<Tamagotchi> {
  return req<Tamagotchi>(token, "POST", "/api/tamagotchi/rename", { name });
}

/**
 * GET /api/tamagotchi/shop
 * Shop items available for the Tamagotchi you control (partner's).
 */
export function fetchShop(token: string): Promise<ShopResponse> {
  return req<ShopResponse>(token, "GET", "/api/tamagotchi/shop");
}

/**
 * POST /api/tamagotchi/shop/buy
 * Purchase an item; XP is deducted from the Tamagotchi you control.
 */
export function buyItem(token: string, itemId: string): Promise<TamagotchiState> {
  return req<TamagotchiState>(token, "POST", "/api/tamagotchi/shop/buy", {
    item_id: itemId,
  });
}

/**
 * POST /api/tamagotchi/equip
 * Equip an owned item to a slot on the Tamagotchi you control.
 */
export function equipItem(
  token: string,
  itemId: string,
  slot: EquippedSlot,
): Promise<TamagotchiState> {
  return req<TamagotchiState>(token, "POST", "/api/tamagotchi/equip", {
    item_id: itemId,
    slot,
  });
}

/**
 * GET /api/tamagotchi/events
 * Recent events for both Tamagotchis in the couple (activity feed).
 */
export function fetchEvents(token: string): Promise<EventsResponse> {
  return req<EventsResponse>(token, "GET", "/api/tamagotchi/events");
}

/**
 * PATCH /api/tamagotchi/appearance
 * Updates free cosmetic fields on YOUR Tamagotchi (species, background,
 * animation, position). These are NOT shop-gated.
 * Only the fields you pass are updated — omit a field to leave it unchanged.
 */
export function updateAppearance(
  token: string,
  patch: {
    species?: string;
    background?: string;
    animation?: string;
    position?: string;
  },
): Promise<TamagotchiState> {
  return req<TamagotchiState>(token, "PATCH", "/api/tamagotchi/appearance", patch);
}

// ─── Derived helpers ──────────────────────────────────────────────────────────

/** XP progress within the current level, as a 0–1 fraction. */
export function xpProgress(state: TamagotchiState): number {
  const { xp_this_level, xp_next_level } = state as unknown as ShopResponse;
  if (!xp_next_level) return 1;
  return Math.min(xp_this_level / xp_next_level, 1);
}

/** Health as a 0–1 fraction. */
export function healthFraction(tama: Tamagotchi): number {
  if (!tama.max_health) return 0;
  return Math.min(tama.health / tama.max_health, 1);
}

/** Find what's currently equipped in a given slot. */
export function equippedInSlot(
  equipped: TamagotchiEquipped[],
  slot: EquippedSlot,
): TamagotchiItem | undefined {
  return equipped.find((e) => e.slot === slot)?.item;
}

/** Returns true if the item is in the inventory. */
export function isOwned(
  inventory: TamagotchiInventoryEntry[],
  itemId: string,
): boolean {
  return inventory.some((e) => e.item.id === itemId);
}